src/sync/rpc.ts

```diff
@@ -0,0 +1,292 @@
+/**
+ * @module sync
+ */
+
+import assert from "assert";
+import BN from "bn.js";
+import {hashTreeRoot} from "@chainsafe/ssz";
+import {
+  bytes32, Slot, number64,
+  BeaconBlockHeader, BeaconBlockBody,
+  RequestBody, Hello, Goodbye, Status,
+  BeaconBlockRootsRequest, BeaconBlockRootsResponse,
+  BeaconBlockHeadersRequest, BeaconBlockHeadersResponse,
+  BeaconBlockBodiesRequest, BeaconBlockBodiesResponse,
+  BeaconStatesRequest, BeaconStatesResponse, Epoch,
+} from "../types";
+import {ZERO_HASH, Method, RequestId, ResponseCode} from "../constants";
+import logger from "../logger";
+import {intDiv} from "../util/math";
+import {IBeaconDb} from "../db";
+import {IBeaconChain} from "../chain";
+import {INetwork} from "../network";
+import {getEmptyBlockBody} from "../chain/genesis";
+import {ReputationStore} from "./reputation";
+
+// eslint-disable-next-line @typescript-eslint/no-empty-interface
+interface SyncOptions {
+}
+
+/**
+ * The SyncRpc module handles app-level requests / responses from other peers,
+ * fetching state from the chain and database as needed.
+ */
+export class SyncRpc {
+  private opts: SyncOptions;
+  private db: IBeaconDb;
+  private chain: IBeaconChain;
+  private network: INetwork;
+  private reps: ReputationStore;
+  public constructor(opts: SyncOptions, {db, chain, network, reps}) {
+    this.opts = opts;
+    this.db = db;
+    this.chain = chain;
+    this.network = network;
+    this.reps = reps;
+  }
+
+  // create common requests
+
+  public async createHello(): Promise<Hello> {
+    let bestSlot: Slot,
+      bestRoot: bytes32,
+      latestFinalizedEpoch: Epoch,
+      latestFinalizedRoot: bytes32;
+    if (!this.chain.genesisTime) {
+      bestSlot = 0;
+      bestRoot = ZERO_HASH;
+      latestFinalizedEpoch = 0;
+      latestFinalizedRoot = ZERO_HASH;
+    } else {
+      bestSlot = await this.db.getChainHeadSlot();
+      bestRoot = await this.db.getBlockRoot(bestSlot);
+      const state = await this.db.getState();
```

Comment:
those two can be promise.all. It somewhat important as this method would probably get called a lot